src/ck3/data/artifacts.rs

```diff
@@ -257,3 +258,45 @@ impl DbKind for ArtifactFeatureGroup {
         let mut _vd = Validator::new(block, data);
     }
 }
+
+#[derive(Clone, Debug)]
+pub struct ArtifactBlueprint {}
+
+inventory::submit! {
+    ItemLoader::Normal(GameFlags::Ck3, Item::ArtifactBlueprint, ArtifactBlueprint::add)
+}
+
+impl ArtifactBlueprint {
+    pub fn add(db: &mut Db, key: Token, block: Block) {
+        db.add(Item::ArtifactBlueprint, key, block, Box::new(Self {}));
+    }
+}
+
+impl DbKind for ArtifactBlueprint {
+    fn validate(&self, _key: &Token, block: &Block, data: &Everything) {
+        let mut vd = Validator::new(block, data);
+
+        vd.req_field("in_type");
+        vd.req_field("in_visuals");
+        vd.field_item("in_type", Item::ArtifactType);
+        vd.field_item("in_visuals", Item::ArtifactVisual);
+
+        vd.req_field("out_type");
+        vd.req_field("out_visuals");
+        vd.field_item("out_type", Item::ArtifactType);
+        vd.field_item("out_visuals", Item::ArtifactVisual);
+
+        vd.field_validated_list("disallowed_modifiers", |token, data| {
+            verify_modif_exists(token, data, ModifKinds::all(), Severity::Warning);
+        });
+        vd.field_validated_block("replacement_modifiers", |block, data| {
+            let mut vd = Validator::new(block, data);
+            for field in ARTIFACT_RARITY {
+                // There seems to be no way to verify that these are specifically artifact modifiers
+                vd.field_list_items(field, Item::Modifier);
```

Comment:
Again, if possible, please verify that these will work on Character

Reply:
Yeah that can even be a general upgrade. Modifier's `validate_call` can deduce the `ModifKind` from the scope types of the received `sc` and verify that all the modifer's modifs apply to that modifkind.

Reply:
It turned out that `validate_call` doesn't need to do that because the callers already call `validate_property_use` :) So I made the artifact code call that too.